Add `simple_subtype_field_names` `NameProposer`

supersaiyansubtlety opened the pull request:
Adds the option of naming simple types' subtypes based on a transformation of the super simple type's name.

Heavily based on the simple types proposer.

example use case: `Entity` subclasses with suffix `Entity` -> `PiglinEntity` fields and params are named `piglin`

---

### Additional fixes

-  `SimpleTypeSingleIndex`:
    - used lower case names for constant conflict detection
    - could give outer class fields an inner class as a parent (no known consequences)
    - inheritance ancestor order iteration was not consistent; inheritance overrides didn't always work
- `ConflictFixProposer` looked up mappings using deobf params, breaking it for deobf method param conflicts which can arise from `DelegateParametersNameProposer`
- simple type verification test flakiness; separates it into its own test input source set and catches thrown exception instead of capturing log output

---

### Format

Uses the same registry as simple types and expands `"inherit"` to allow object values. Inherit can now take any of the following forms (also see schema):

~~`"inherit": true`~~ removed
```json5
"inherit": {
  // equivalent to old "inherit": true
  "type": "DIRECT"
}
```
```json5
"inherit": {
  // names subclass fields by checking if their type ends with the suffix, then names them after their type with the suffix removed
  "type": "TRUNCATED_SUBTYPE_NAME",
  "suffix": "Entity"
}
```
```json5
"inherit": {
  // names subclass fields by checking if their type matches the pattern, then names them after their type with the replacement applied
  "type": "TRANSFORMED_SUBTYPE_NAME",
  "pattern": "(.+)BlockEntityRenderer",
  "replacement": "$1Renderer"
}
```

---

Todo:
- [x] don't do subtype naming for abstract classes (so `LivingEntity`s aren't named `living`)
- [x] more tests

supersaiyansubtlety committed:
initial working version of simple subtypes

OroArmor reviewed (changes requested):
I think this would be better as an extension to the simple type field name that already exists, with maybe `"inherit_suffix": "Entity"` or something.

supersaiyansubtlety commented:
> I think this would be better as an extension to the simple type field name that already exists, with maybe `"inherit_suffix": "Entity"` or something.

I'm not sure how that would work.
This proposer is dynamic because it's based on already mapped class names. I don't see how that could be combined with existing simple type names.

supersaiyansubtlety committed:
checkstyle

supersaiyansubtlety commented:
I guess they *could* be combined and just be doing effectively two separate things.
I don't see a technical reason to combine them though.

supersaiyansubtlety committed:
merge SimpleSubtypeFieldNamesRegistry into SimpleTypeFieldNamesRegistry
implement regex replacement subtype proposer
fix SimpleTypeSingleIndex bug that used constant's local names for conflict detection
fix SimpleTypeSingleIndex bug that could give outer class fields an inner class as a parent (no known consequences)

supersaiyansubtlety committed:
improve SimpleTypeFieldNamesRegistry validation and schema
log errors instead of throwing during "inherit" parsing

supersaiyansubtlety committed:
make SimpleSubtypeSingleIndex skip non-concrete types

supersaiyansubtlety committed:
add tests for TRANSFORM_SUBTYPE_NAME
add tests for duplicate subtype fields+locals

supersaiyansubtlety committed:
add clarifying test comment

supersaiyansubtlety commented:
Looking into an issue I found by trying the `Entity` subtype on MC mappings

supersaiyansubtlety committed:
fix merging of outerclass fields in SimpleSubtypeSingleIndex

supersaiyansubtlety committed:
make simple subtype proposals incremental

supersaiyansubtlety committed:
remove unnecessary type fields

supersaiyansubtlety committed:
fix for packaged classes

supersaiyansubtlety committed:
Merge branch 'master' into simple-subtypes
extract SimpleSubtypeFieldNameProposerTest
add TODO for failing simple subtype duplicate test
add TODO for flaky simple type verification test

supersaiyansubtlety committed:
fix starting param type counts at 0 instead of 1 for SimpleSubtypeSingleIndex conflict prevention

supersaiyansubtlety committed:
fix ConflictFixProposer for deobf methods' params

supersaiyansubtlety committed:
fix type verification test flakiness with polling hack

supersaiyansubtlety committed:
minor cleanup

supersaiyansubtlety committed:
reduce visibility of some simple sub/type index members

supersaiyansubtlety committed:
remove unused @TestOnly method

supersaiyansubtlety committed:
only read SimpleTypeFieldNamesRegistry once

supersaiyansubtlety committed:
split up SimpleSubtypeFieldnameProposerTest tests

supersaiyansubtlety committed:
inline sub/type indexes' dropCache methods

supersaiyansubtlety committed:
fix inheritance search order in SimpleSubtypeSingleIndex
add tests for overriding inheritted subtype rules
move+fix isValidJavaIdentifier check for SimpleSubtypeFieldNameProposer
add tests for not proposing invalid subtype names

supersaiyansubtlety committed:
replace insertAndDynamicallyPropose with putMapping

supersaiyansubtlety committed:
fix inheritance overrides for simple type names

supersaiyansubtlety commented:
I've made substantial changes, so new diffs:
[entity-qmap.diff.txt](https://github.com/user-attachments/files/23844514/entity-qmap.diff.txt)
[ber-qmap.diff.txt](https://github.com/user-attachments/files/23844466/ber-qmap.diff.txt)

supersaiyansubtlety committed:
add separate simple type verification test

supersaiyansubtlety committed:
remove SimpleTypeFieldNameProposerTest::testTypeVerification
set type verification to THROW for simple sub/type test input

ix0rai commented on src/main/java/org/quiltmc/enigma_plugin/proposal/SimpleSubtypeFieldNameProposer.java:
```diff
@@ -0,0 +1,109 @@
+	public static final String ID = "simple_subtype_field_names";
+	private final SimpleSubtypeSingleIndex index;
+
+	private static String unCapitalize(String string) {
```
why not just use toLowerCase(Local.ROOT)? i imagine there's slightly less overhead here, but it's so negligeable

supersaiyansubtlety replied:
`unCapitalize` only converts the first character, `toLowerCase` converts all characters
`unCapitalize` is for converting from `UpperCamelCase` to `lowerCamelCase`

ix0rai replied:
👍 oops

ix0rai commented on src/main/java/org/quiltmc/enigma_plugin/index/simple_type_single/SimpleTypeFieldNamesRegistry.java:
```diff
@@ -182,6 +232,145 @@ public Entry(String type, String localName, String staticName) {
+
+		static Either<? extends Inherit, String> read(JsonReader reader) throws IOException {
+			switch (reader.peek()) {
+				case BOOLEAN -> {
```
i'd prefer the boolean logic was removed in favor of directly naming the scheme, i think that inherit: true has always been confusing and i'd like to entirely remove it

eristhea reviewed (changes requested):
looks great, this feature is really well done! just wondering why you reimplemented toLowerCase() lol

supersaiyansubtlety committed:
unify unCapitalize methods

supersaiyansubtlety committed:
replace Either with Result for SimpleTypeFieldNamesRegistry parsing

supersaiyansubtlety committed:
remove boolean parsing for inherit property

supersaiyansubtlety committed:
extract missingRequirementErrOf method

supersaiyansubtlety committed:
Merge branch 'master' into simple-subtypes

supersaiyansubtlety committed:
update to enigma 2.7.2, use InheritanceIndex::streamAncestors

supersaiyansubtlety commented on src/main/java/org/quiltmc/enigma_plugin/util/EntryUtil.java:
```diff
@@ -61,4 +65,17 @@ public static String getNonHashedNameOrNull(Entry<?> entry) {
+	 * @return a breadth-first stream of the passed {@code classEntry}'s ancestors;
+	 * order within one level of depth is not guaranteed
+	 */
+	public static Stream<ClassEntry> streamAncestors(ClassEntry classEntry, InheritanceIndex inheritance) {
```
347 was part of the Enigma 2.7.2 release and its version of `streamAncestors` fixes this

supersaiyansubtlety commented:
ready!

eristhea reviewed (changes requested):
we should mark this as only compatible with `2.7.2+` since we're using the new API!

supersaiyansubtlety commented:
> we should mark this as only compatible with `2.7.2+` since we're using the new API!

good catch

supersaiyansubtlety committed:
require enigma 2.7.2

eristhea reviewed (approved):
lgtm!
